Makefile

```diff
@@ -7,25 +7,26 @@ ifneq (,$(wildcard /proc/sys/fs/binfmt_misc/WSLInterop))
 else
   JAR_PATH = /usr/share/kaitai-struct-compiler/lib/*
 endif
-all: src-animal/libanimal.so
+KSY := animal fake
+LIBS := $(foreach ksy,$(KSY),test_artifacts/lib$(ksy).so)
 
-src-animal/libanimal.so: $(BUILD) src-animal/animal.cpp
-	PYTHONPATH=$$PYTHONPATH:local $(BUILD) src-animal/animal.cpp -b build
+test_artifacts/lib%.so: test_artifacts/%.cpp $(BUILD)
```

Comment:
@ManasviGoyal I am putting the source and compiled files for the tests all in a single folder to simplify the pattern matching in the Makefile.

Unfortunately this causes a compilation issue:

```
[100%] Linking CXX shared library libfake.so
/usr/bin/ld: CMakeFiles/fake.dir/home/runner/work/kaitai_struct_awkward_runtime/kaitai_struct_awkward_runtime/test_artifacts/fake.cpp.o:(.bss+0x0): multiple definition of `builder_map[abi:cxx11]'; CMakeFiles/fake.dir/home/runner/work/kaitai_struct_awkward_runtime/kaitai_struct_awkward_runtime/test_artifacts/animal.cpp.o:(.bss+0x0): first defined here
/usr/bin/ld: CMakeFiles/fake.dir/home/runner/work/kaitai_struct_awkward_runtime/kaitai_struct_awkward_runtime/test_artifacts/fake.cpp.o:(.bss+0x30): multiple definition of `builder_keys[abi:cxx11]'; CMakeFiles/fake.dir/home/runner/work/kaitai_struct_awkward_runtime/kaitai_struct_awkward_runtime/test_artifacts/animal.cpp.o:(.bss+0x30): first defined here
/usr/bin/ld: CMakeFiles/fake.dir/home/runner/work/kaitai_struct_awkward_runtime/kaitai_struct_awkward_runtime/test_artifacts/fake.cpp.o: in function `load(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)':
fake.cpp:(.text+0x679): multiple definition of `load(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)'; CMakeFiles/fake.dir/home/runner/work/kaitai_struct_awkward_runtime/kaitai_struct_awkward_runtime/test_artifacts/animal.cpp.o:animal.cpp:(.text+0x945): first defined here
/usr/bin/ld: CMakeFiles/fake.dir/home/runner/work/kaitai_struct_awkward_runtime/kaitai_struct_awkward_runtime/test_artifacts/fake.cpp.o: in function `fill':
fake.cpp:(.text+0x816): multiple definition of `fill'; CMakeFiles/fake.dir/home/runner/work/kaitai_struct_awkward_runtime/kaitai_struct_awkward_runtime/test_artifacts/animal.cpp.o:animal.cpp:(.text+0xae2): first defined here
/usr/bin/ld: CMakeFiles/fake.dir/home/runner/work/kaitai_struct_awkward_runtime/kaitai_struct_awkward_runtime/test_artifacts/fake.cpp.o: in function `form':
fake.cpp:(.text+0x99c): multiple definition of `form'; CMakeFiles/fake.dir/home/runner/work/kaitai_struct_awkward_runtime/kaitai_struct_awkward_runtime/test_artifacts/animal.cpp.o:animal.cpp:(.text+0xc68): first defined here
/usr/bin/ld: CMakeFiles/fake.dir/home/runner/work/kaitai_struct_awkward_runtime/kaitai_struct_awkward_runtime/test_artifacts/fake.cpp.o: in function `length':
fake.cpp:(.text+0xa48): multiple definition of `length'; CMakeFiles/fake.dir/home/runner/work/kaitai_struct_awkward_runtime/kaitai_struct_awkward_runtime/test_artifacts/animal.cpp.o:animal.cpp:(.text+0xd14): first defined here
/usr/bin/ld: CMakeFiles/fake.dir/home/runner/work/kaitai_struct_awkward_runtime/kaitai_struct_awkward_runtime/test_artifacts/fake.cpp.o: in function `num_buffers':
fake.cpp:(.text+0xa66): multiple definition of `num_buffers'; CMakeFiles/fake.dir/home/runner/work/kaitai_struct_awkward_runtime/kaitai_struct_awkward_runtime/test_artifacts/animal.cpp.o:animal.cpp:(.text+0xd32): first defined here
/usr/bin/ld: CMakeFiles/fake.dir/home/runner/work/kaitai_struct_awkward_runtime/kaitai_struct_awkward_runtime/test_artifacts/fake.cpp.o: in function `buffer_name':
fake.cpp:(.text+0xa84): multiple definition of `buffer_name'; CMakeFiles/fake.dir/home/runner/work/kaitai_struct_awkward_runtime/kaitai_struct_awkward_runtime/test_artifacts/animal.cpp.o:animal.cpp:(.text+0xd50): first defined here
/usr/bin/ld: CMakeFiles/fake.dir/home/runner/work/kaitai_struct_awkward_runtime/kaitai_struct_awkward_runtime/test_artifacts/fake.cpp.o: in function `buffer_size':
fake.cpp:(.text+0xb43): multiple definition of `buffer_size'; CMakeFiles/fake.dir/home/runner/work/kaitai_struct_awkward_runtime/kaitai_struct_awkward_runtime/test_artifacts/animal.cpp.o:animal.cpp:(.text+0xe0f): first defined here
/usr/bin/ld: CMakeFiles/fake.dir/home/runner/work/kaitai_struct_awkward_runtime/kaitai_struct_awkward_runtime/test_artifacts/fake.cpp.o: in function `copy_into':
fake.cpp:(.text+0xbbc): multiple definition of `copy_into'; CMakeFiles/fake.dir/home/runner/work/kaitai_struct_awkward_runtime/kaitai_struct_awkward_runtime/test_artifacts/animal.cpp.o:animal.cpp:(.text+0xe88): first defined here
/usr/bin/ld: CMakeFiles/fake.dir/home/runner/work/kaitai_struct_awkward_runtime/kaitai_struct_awkward_runtime/test_artifacts/fake.cpp.o: in function `deallocate':
fake.cpp:(.text+0xbf2): multiple definition of `deallocate'; CMakeFiles/fake.dir/home/runner/work/kaitai_struct_awkward_runtime/kaitai_struct_awkward_runtime/test_artifacts/animal.cpp.o:animal.cpp:(.text+0xebe): first defined here
collect2: error: ld returned 1 exit status
gmake[3]: *** [CMakeFiles/fake.dir/build.make:116: libfake.so] Error 1
gmake[3]: Leaving directory '/home/runner/work/kaitai_struct_awkward_runtime/kaitai_struct_awkward_runtime/build/build'
gmake[2]: *** [CMakeFiles/Makefile2:115: CMakeFiles/fake.dir/all] Error 2
gmake[2]: Leaving directory '/home/runner/work/kaitai_struct_awkward_runtime/kaitai_struct_awkward_runtime/build/build'
gmake[1]: *** [Makefile:136: all] Error 2
gmake[1]: Leaving directory '/home/runner/work/kaitai_struct_awkward_runtime/kaitai_struct_awkward_runtime/build/build'
Traceback (most recent call last):
  File "/home/runner/work/kaitai_struct_awkward_runtime/kaitai_struct_awkward_runtime/local/bin/awkward-kaitai-build", line 8, in <module>
    sys.exit(main())
  File "/home/runner/work/kaitai_struct_awkward_runtime/kaitai_struct_awkward_runtime/local/awkward_kaitai/builder.py", line 154, in main
    build_with_cmake(args.file, dest, args.build)
  File "/home/runner/work/kaitai_struct_awkward_runtime/kaitai_struct_awkward_runtime/local/awkward_kaitai/builder.py", line 124, in build_with_cmake
    subprocess.run(
  File "/opt/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/subprocess.py", line 526, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '[PosixPath('/usr/local/bin/cmake'), '--build', 'build/build']' returned non-zero exit status 2.
make: *** [Makefile:14: test_artifacts/libfake.so] Error 1
rm test_artifacts/fake.cpp test_artifacts/animal.cpp
Error: Process completed with exit code 2.
```

See https://github.com/ManasviGoyal/kaitai_struct_awkward_runtime/actions/runs/7457486232/job/20289758213#step:7:600

It seems like the `CMake` step mixes up the 2 libraries, is there a way to fix this?
What is strange is that if I run the compilation step twice, it actually works the second time.

Reply:
@ManasviGoyal the fix on-the-fly during the call was amazing, thanks!